llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -2921,210 +2921,137 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpAccessChain>(SPIRVValue *
+  Value *base = transValue(spvAccessChain->getBase(), getBuilder()->GetInsertBlock()->getParent(),
+                           getBuilder()->GetInsertBlock());
+  auto srcIndices = transValue(spvAccessChain->getIndices(), getBuilder()->GetInsertBlock()->getParent(),
+                               getBuilder()->GetInsertBlock());
 
-  truncConstantIndex(indices, getBuilder()->GetInsertBlock());
+  truncConstantIndex(srcIndices, getBuilder()->GetInsertBlock());
 
   if (!spvAccessChain->hasPtrIndex())
-    indices.insert(indices.begin(), getBuilder()->getInt32(0));
+    srcIndices.insert(srcIndices.begin(), getBuilder()->getInt32(0));
 
-  SPIRVType *const spvBaseType = spvAccessChain->getBase()->getType();
+  SPIRVType *spvAccessType = spvAccessChain->getBase()->getType();
+  const SPIRVStorageClassKind storageClass = spvAccessType->getPointerStorageClass();
+  const bool isBufferBlockPointer = isStorageClassExplicitlyLaidOut(m_bm, storageClass);
 
   Type *basePointeeType = getPointeeType(spvAccessChain->getBase());
   // TODO: Remove this when LLPC will switch fully to opaque pointers.
   assert(IS_OPAQUE_OR_POINTEE_TYPE_MATCHES(base->getType(), basePointeeType));
 
-  SPIRVType *spvAccessType = spvBaseType;
-
-  // Records where (if at all) we have to split our indices - only required when going through a row_major matrix or
-  // if we indexing into a struct that has partially overlapping offsets (normally occurs with HLSL cbuffer packing).
-  SmallVector<std::pair<unsigned, Type *>, 4> splits;
+  SmallVector<Value *, 8> gepIndices;
 
-  const SPIRVStorageClassKind storageClass = spvBaseType->getPointerStorageClass();
+  if (spvAccessType->isTypeForwardPointer())
+    spvAccessType = static_cast<SPIRVTypeForwardPointer *>(spvAccessType)->getPointer();
+  assert(spvAccessType->isTypePointer());
+  gepIndices.push_back(srcIndices[0]);
+  spvAccessType = spvAccessType->getPointerElementType();
 
-  const bool isBufferBlockPointer = isStorageClassExplicitlyLaidOut(m_bm, storageClass);
+  auto flushGep = [&]() {
+    if (gepIndices.size() == 1) {
+      if (auto *constant = dyn_cast<ConstantInt>(gepIndices[0])) {
+        if (constant->getZExtValue() == 0)
+          return; // no-op
+      }
+    }
 
-  // Run over the indices of the loop and investigate whether we need to add any additional indices so that we load
-  // the correct data. We explicitly lay out our data in memory, which means because Vulkan has more powerful layout
-  // options to producers than LLVM can model, we have had to insert manual padding into LLVM types to model this.
-  // This loop will ensure that all padding is skipped in indexing.
-  for (unsigned i = 0; i < indices.size(); i++) {
-    bool isDone = false;
+    if (spvAccessChain->isInBounds())
+      base = getBuilder()->CreateInBoundsGEP(basePointeeType, base, gepIndices);
+    else
+      base = getBuilder()->CreateGEP(basePointeeType, base, gepIndices);
```

Comment:
Not new code, but this could also be
```c++
base = getBuilder()->CreateGEP(basePointeeType, base, gepIndices, "", spvAccessChain->isInBounds());
```